src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,122 @@
+             (constant-node-p env node constant-bindings))
+
+           (propagate-constants-node-let (node constant-bindings)
+             (declare (type node-let node))
+             (let ((node-bindings (node-let-bindings node))
+                   (new-constant-bindings nil)
+                   (nonconstant-bindings nil))
+               (loop :for (var . value) :in node-bindings
+                     :for propagated-value-node := (funcall *traverse* value constant-bindings)
+                     :do (if (constant-node-p env propagated-value-node constant-bindings)
+                             (push (cons var propagated-value-node)
+                                   new-constant-bindings)
+                             (push (cons var propagated-value-node)
+                                   nonconstant-bindings)))
+               (let ((inner-constant-bindings (append new-constant-bindings constant-bindings)))
+                 (if nonconstant-bindings
+                     (make-node-let
+                      :type (node-type node)
+                      :bindings nonconstant-bindings
+                      :subexpr (funcall *traverse* (node-let-subexpr node) inner-constant-bindings))
+                     (funcall *traverse* (node-let-subexpr node) inner-constant-bindings)))))
+
+           (propagate-constants-node-lisp (node constant-bindings)
+             (declare (type node-lisp node))
+             (let* ((new-lisp-vars nil)
+                    (let-bindings
+                      (loop :for (lisp-var . coalton-var) :in (node-lisp-vars node)
+                            :for constant-value := (constant-var-value coalton-var constant-bindings :no-error t)
+                            :if constant-value
+                              :collect (let ((new-coalton-var (gentemp (symbol-name coalton-var))))
+                                         (push (cons lisp-var new-coalton-var) new-lisp-vars)
+                                         (cons new-coalton-var constant-value))
+                            :else
+                              :do (push (cons lisp-var coalton-var) new-lisp-vars))))
+               (if let-bindings
+                   (make-node-let
+                    :type (node-type node)
+                    :bindings let-bindings
+                    :subexpr (make-node-lisp
+                              :type (node-type node)
+                              :vars new-lisp-vars
+                              :form (node-lisp-form node)))
+                   (make-node-lisp
+                    :type (node-type node)
+                    :vars new-lisp-vars
+                    :form (node-lisp-form node)))))
+
+           (direct-application-better-infer-types (node constant-bindings)
+             (declare (type node-direct-application node)
+                      (ignore constant-bindings))
+             ;; FIXME: Can we do something similar for non-direct APPLICATION?
```

Comment:
Just saw that `node-application` does not have a `rator-type`. Then, we can straightaway remove this FIXME?

```suggestion
```

Reply:
I mean, the `node-application` still has a `rator` which is a node which has a `type`, but I don't know whether anything would be gained from trying to rewrite its type based on any operand types

Reply:
I mean if you look at the definitions
```lisp
(defstruct (node-variable (:include node))
  "Variables like x or y"
  (value (util:required 'value) :type parser:identifier :read-only t))

(defstruct (node-application (:include node))
  "Function application (f x)"
  (rator (util:required 'rator) :type node      :read-only t)
  (rands (util:required 'rands) :type node-list :read-only t))

(defstruct (node-direct-application (:include node))
  "Fully saturated function application of a known function"
  (rator-type (util:required 'rator-type) :type tc:ty             :read-only t)
  (rator      (util:required 'rator)      :type parser:identifier :read-only t)
  (rands      (util:required 'rands)      :type node-list         :read-only t))
```
you see that the `node-application`'s `rator` is an actual `node`, while the `node-direct-application`'s `rator` is just a `parser:identifier`, so it needs another field `rator-type` to specify the corresponding type. It would probably be an equivalent idea to write
```lisp
(defstruct (node-direct-application (:include node))
  "Fully saturated function application of a known function"
  (rator (util:required 'rator) :type node-variable :read-only t)
  (rands (util:required 'rands) :type node-list     :read-only t))
```
since this still requires that the `rator` is a "known function" / named variable. I imagine the reason for flattening it in the current `node-direct-application` is just for accessor convenience.

Reply:
Since the `rator` is a `node` for the `node-application`, can we expect that the default traverse would update the types?

Reply:
No, as I mentioned in another comment, the default traversal only traverses through nodes, leaving types unchanged (see `src/codegen/traverse.lisp`:
```lisp
    (action (:traverse node-application node &rest args)
      (make-node-application
       :type (node-type node)
       :rator (apply *traverse* (node-application-rator node) args)
       :rands (mapcar
               (lambda (node)
                 (apply *traverse* node args))
               (node-application-rands node))))
    (action (:traverse node-direct-application node &rest args)
      (make-node-direct-application
       :type (node-type node)
       :rator-type (node-direct-application-rator-type node)
       :rator (node-direct-application-rator node)
       :rands (mapcar
               (lambda (node)
                 (apply *traverse* node args))
               (node-direct-application-rands node))))
```
If you want to change the types of applications' `rator`s to always match the operands, you could rewrite these to update them, or you could also just redefine the function `node-rator-type` in `ast.lisp` to compute the type based on the arguments every time it gets called

Reply:
I think I asked the wrong question. I should have asked only about the `rator` not about its type. For `node-application`, we do not need to do anything more than what is currently being done by the default traverse functions to update the `rator`. 

For `node-direct-application`, we decided for the PR-specific solution.

I am now having a feeling that type updation can also be carried out for node-let, and probably for several other nodes. A PR incorporating this will also update the type of `rator` for `node-application`.

I also noted that `direct-application-better-infer-types` is never called while loading coalton-library even if heuristic inlining is enabled. However, it gets called during monomorphization, such as below:

```lisp
(coalton-toplevel
  (define (partial-monomorphization-a x y z)
    (== x y)
    (+ 1 z)))

(coalton-toplevel
  (declare partial-monomorphization-b (Eq :a => :a -> :a -> Integer -> Integer))
  (monomorphize)
  (define (partial-monomorphization-b x y z)
    (partial-monomorphization-a x y z)))
```

Reply:
Yes, I think a `node-direct-application` only gets _created_ when calling `src/codegen/optimizer.lisp`'s `direct-application` function. A search shows just two times where `direct-application` is called:

- inside of `src/entry.lisp`'s `expression-entry-point`:
```lisp
...
            (when (null preds)
              (return-from expression-entry-point
                (let ((node (codegen:optimize-node
                             (codegen:translate-expression node nil env)
                             env)))
                  (codegen:codegen-expression 
                   (codegen:direct-application
                    node
                    (codegen:make-function-table env))
                   env))))
...
```

- inside `src/codegen/optimizer.lisp`'s `optimize-bindings`:
```lisp
...
  (let ((bindings (optimize-bindings-initial bindings package env)))
...
      (setf bindings
            (loop :for (name . node) :in bindings
                  :if (gethash name monomorphize-table)
                    :append (optimize-bindings-initial
                             (monomorphize
                              name
                              manager
                              package
                              resolve-table
                              (lambda (node env)
                                (optimize-node node env))
                              env)
                             package env)
                  :else
                    :collect (cons name node)))
...
        (setf bindings
              (loop :for (name . node) :in bindings
                    :collect (cons name (direct-application node function-table))))
...
```

In both of these cases, `optimize-node` (and/or `optimize-bindings-initial`) gets called _before_ `direct-application`. So I guess it's only possible for the functions in `optimize-node` to receive a `node-direct-application` if a recursive call to the optimizer has already been completed, like how the monomorphizer recursively calls back to the optimizer?

Reply:
Ah, yes indeed! This point about when `node-direct-application` are created had also risen in earlier discussions with stylewarning and eliaslfox. Your elaboration pretty much sums it up.

But coming back to resolving this. I see two steps -

1. For this PR, leave both node-application and node-direct-application untouched, unless it directly concerns constant propagation.
2. In a separate PR, incorporate type updation in the default traverse functions.